code/controllers/subsystem/lighting.dm

```diff
@@ -32,18 +32,22 @@ SUBSYSTEM_DEF(lighting)
 
 	var/list/queue = sources_queue
 	var/i = 0
-	for (i in 1 to length(queue))
-		var/datum/light_source/L = queue[i]
+	while(i < length(queue)) //we don't use for loop here because i cannot be changed during an iteration
+		i += 1
 
+		var/datum/light_source/L = queue[i]
 		L.update_corners()
-		L.needs_update = LIGHTING_NO_UPDATE
+		if(!QDELETED(L))
+			L.needs_update = LIGHTING_NO_UPDATE
+		else
+			i -= 1 // update_corners() has removed L from the list, move back so we don't overflow or skip the next element
 
 		if(init_tick_checks)
 			CHECK_TICK
```

Comment:
```suggestion
		// We unroll TICK_CHECK here so we can clear out the queue to ensure any removals/additions when sleeping don't fuck us
		if(init_tick_checks)
			if(!TICK_CHECK)
				continue
			queue.Cut(1, i+1)
			i = 0
			stoplag()
			queue_length = length(queue)
```
The risk of being removed while iterating also applies while sleeping, and because we don't know when or how many removals happen, we risk double processing or failing to process some entries, so fully clearing the queue of processed items before sleeping is necessary

Reply:
Don't need `queue_length` the while loop condition checks that for us